src/http-gateways/trustless-gateway.md

```diff
@@ -66,12 +74,146 @@ Below response types MUST to be supported:
+said entity.
+
+Allowed values for `from` and `to` are positive integers where `to` >= `from`, which
+limit the return blocks to needed to satisfy the range `[from,to]`:
+
+- `from` value gives the byte-offset of the first byte in a range.
+- `to` value gives the byte-offset of the last byte in the range; that is,
+the byte positions specified are inclusive.  Byte offsets start at zero.
+
+If the entity at the end of the path cannot be interpreted as a continuous
+array of bytes (such as a DAG-CBOR/JSON map, or UnixFS directory), this
+parameter has no effect.
+
+The following additional values are supported:
+
+- `*` can be substituted for end-of-file
+  - `entity-bytes=0:*` is the entire file (a verifiable version of HTTP request for `Range: 0-`)
+- Negative numbers can be used for referring to bytes from the end of a file
+  - `entity-bytes=-1024:*` is the last 1024 bytes of a file
+    (verifiable version of HTTP request for `Range: -1024`)
+  - It is also permissible (unlike with HTTP Range Requests) to ask for the
+    range of 500 bytes from the beginning of the file to 1000 bytes from the
+    end: `entity-bytes=499:-1000`
+
+When present, returned `Etag` must include unique prefix based on the passed range.
+
 # HTTP Response
 
 Below MUST be implemented **in addition** to "HTTP Response" of :cite[path-gateway].
 
 ## HTTP Response Headers
 
+### `Content-Type` (response header)
+
+MUST be returned and include additional format-specific parameters when possible.
+
+If a CAR stream was requested, the response MUST include the parameter specifying CAR version.
+For example: `Content-Type: application/vnd.ipld.car; version=1`
+
 ### `Content-Disposition` (response header)
 
 MUST be returned and set to `attachment` to ensure requested bytes are not rendered by a web browser.
+
+## HTTP Response Payload
+
+### Block Response
+
+An opaque bytes matching the requested block CID
+([application/vnd.ipld.raw](https://www.iana.org/assignments/media-types/application/vnd.ipld.raw)).
+
+The Body hash MUST match the Multihash from the requested CID.
+
+### CAR Response
+
+A CAR stream
+([application/vnd.ipld.car](https://www.iana.org/assignments/media-types/application/vnd.ipld.car))
+for the requested content type,  path and optional `dag-scope` and `entity-bytes` URL parameters.
+
+:::note
+
+By default, block order in CAR response is not deterministic, blocks can
+be returned in different order, depending on implementation choices (traversal,
+speed at which blocks arrive from the network, etc). An opt-in ordered CAR
+responses MAY be introduced in a future, see [IPIP-412](https://github.com/ipfs/specs/pull/412).
+
+:::
+
+#### CAR version
+
+Value returned in `CarV1Header.version` struct  MUST match the `version`
+parameter returned in `Content-Type` header
+
+#### CAR roots
+
+:::issue
+
+TODO: we need to specify expectations about what should be returned in
+[`CarV1Header.roots`](https://ipld.io/specs/transport/car/carv1/#header).
```

Comment:
Main concern here is that default behavior of many tools (like Kubo) is to  pin every CID from the `CarV1Header.roots` list recursively on `.car` import.

I've added 4 options (https://github.com/ipfs/specs/pull/402/commits/bfb5d3cf1d51225f0748c32e672915169d9eae4f, below) to make it easier for us to decide which one should be part of the spec.

- Options A and C will work with `ipfs dag import` in Kubo, which has `--pin-roots=true` as the implicit default, but we no longer can assume root CIDs are always present.
- Options B and D  will trigger some over-fetch, but are deterministic, always provide meaningful root information:
  - B is probably the least painful and acceptable for real-world use, the worst case it will overfetch the terminating entity, but we won't pin or overfetch parents.
  - D has a nice property of matching `X-Ipfs-Roots`, which we generate anyway, but is the worst case for all past and current  Kubo versions: `ipfs dag import` will fetch entire Wikipedia when a CAR has only one article + blocks for parents, unless we make significant changes to `ipfs dag import` in Kubo.

I am leaning towards B or D (if we are willing to break/cleanup how `ipfs dag import` works)

cc @rvagg @aschmahmann @hacdias

Reply:
Ok, B is the safest one:
- existing gateways (incl. Kubo  0.20) already do this for `?format=car`
- keeping the behavior avoids breaking change for existing users (overfetching may caus OOM/storage/excessive costs)

This unblocks this IPIP without increasing the scope.
We can always add D in future IPIP, when we have CAR format / content path syntax for expressing partial DAGs with parents of content paths.

Reply:
Done: https://github.com/ipfs/specs/pull/402/commits/0639715a6f895c710a682faf00532a21dcb3beff

Reply:
@hannahhoward  any concerns around this requirement? (this is what existing ipfs.io does).

> If the response uses version 1 or 2 of the CAR spec, the [`CarV1Header.roots`](https://ipld.io/specs/transport/car/carv1/#header) field MUST contain the CID of the terminating entity.

Reply:
@lidel can you clarify (both here and in the doc) what you mean by "terminating entity"? Are we talking about the entity at the terminus of the path, or at some terminus of the DAG? I think you mean path terminus but would you mind being clearer about that in the doc please?

Two problems to raise here:

1. If it's anything other than not the requested root, then we can't conform to this spec in Lassie without buffering, which I don't think we want to be doing, we need to get the CAR header out ASAP and not wait around until we've figured out `/where/this/path/ends/and/how/many/sharded/directories/we/have/to/navigate/through/and/collect/blocks/from`. I don't think we'd be the only party not keen to be buffering before we are able to produce a spec-compliant "trustless" CAR (similar issue, though not quite as bad, for the case of being able to validate a CAR in a streaming way).
2. Because of `<history>`, go-car is going to throw up errors in some forms of consuming (perhaps most, I'd have to trace these code paths in v2) zero-root CARs. I explicitly left the zero-roots possibility in the spec and in js-car, but we didn't iron this out of go-car, mainly because at the time it was a requirement of online deals in Filecoin where it was originally being used: https://github.com/ipld/go-car/blob/e09deeb6f2fded9370b8f5199123cb4894120996/car.go#L144-L146, https://github.com/ipld/go-car/blob/e09deeb6f2fded9370b8f5199123cb4894120996/v2/internal/carv1/car.go#L155-L157

Reply:
> I don't think we'd be the only party not keen to be buffering before we are able to produce a spec-compliant "trustless" CAR [..]

Agree, at the end of the day `roots` value is limited to specific use cases, and in HTTP transport they do not matter, client knows what they've asked. Allowing empty `roots` may end up an important latency optimization for trustless clients like [ipfs in chromium](https://blog.ipfs.tech/2023-05-multigateway-chromium-client/), where miliseconds matter for happy eyeballs.

I've relaxed requirement in https://github.com/ipfs/specs/pull/402/commits/56786ff25ed31c027b07f38b6750df00421cdad7, `roots` is now optional,  and clarified terminus + added explicit note about impact on buffering. 

> I explicitly left the zero-roots possibility in the spec and in js-car, but we didn't iron this out of go-car

Agree, "empty car, no roots" should not be an error.

Reply:
Relevant to the zero-roots question: https://github.com/ipld/go-car/issues/26 - https://github.com/ipld/specs/pull/255/files; current suggestion for compact but safe header is just to use `bafkqaaa`. But I do feel like maybe it's time to just remove that len==1 thing entirely, maybe we could do it in the v2/ code paths at least.

Reply:
@lidel as @rvagg said, `go-car` screams if there aren't any roots. I wouldn't be opposed to changing it to allow reading rootless CARs. I just wonder if that's going to break anything within `go-car`. Perhaps we could go for the empty CID.

Note that conformance tests are blocked on this, since they'll fail if we try opening a rootless CAR.

Reply:
@lidel my first response is why would we specify anything about the behavior in roots in this spec.

Reply:
this should be a seperate IPIP if it's important.

Reply:
And I'm not of the view this is important. If I make a trustless request, and you send me a CAR file back that contains all the blocks needed to trustless verify said request, I do NOT feel that I should reject it cause of what is in the roots field of CARs.

I specifically think the Roots field in the current CAR spec is a foot gun cause its meaning is completely unclear. To the extent people are basing behavior off of it, that's not good.

Reply:
my first preference is leave this out of the spec, say we might define it later.

following that, I would say, entity is not good for streaming because you have to traverse the path blocks first to verify it

after that my preference would be EMPTY

after that my preference would be root of the whole dag

my least favorite is to use entity.

Reply:
We are in a pickle with this one :) 

Implementers outside PL should have some rules of thumb written down, and should not guess, because they may guess and implement the thing we like the least.  

@rvagg  We could adjust spec and replace `the field MAY be left empty.` with ``the field MAY  include `bafkqaaa` CID to indicate that roots were not known before returning the CAR stream``. This way we always have a CID, will not error on existing/legacy software, and if such CAR is imported to future Kubo it will only pin that identity CID, will not trigger over-fetch of entire Wikipedia if I've only exported a single article + path. Solves multiple problems.

BUT, if we say it is ok to use fake `bafkqaaa` CID here, then why are we not ok with [using empty block or identity CID as tombstone for indicating the end of car stream](https://www.notion.so/pl-strflt/Streaming-CARs-bfa54196ed0740eea3c23dd70efbe07b#97f0454e9c094c98a76aeac6184cf3e7)? iirc Juan was against that, as it mixes data layer with transport layer semantics, and unsure how this is different -- I suspect he will be against this one too, for consistency, but maybe the pragmatism will prevail here?

Reply:
The biggest issue here is not one of our making: the roots field of the CAR has unresolved issues, documented in the spec. It's not clear what it's truly supposed to mean in relation to the blocks in the file. (i.e. for example, whether the Root CID even needs to represent an actual block in the collection that follows)

In practice, we have moved farther and farther down the road of treating the CAR as a packed/streaming block store. IMHO, this makes sense cause the CAR format to date doesn't actually contain enough information to treat it as a self-verifying serialization of a DAG traversal. I agree `bafkqaaa` is a weird hack. IMHO, we need to update go-car to handle empty roots. If that requires checking downstream for breaking changes, let's do that. I don't believe the surface area of code that is affected is actually that wide.

Reply:
Anyway, I like where you ended @lidel (MAY contain)

Reply:
Ended up removing this from the scope of this IPIP: https://github.com/ipfs/specs/pull/402#issuecomment-1599428849